timlee/networking-with-a-raspberry/README.md

```diff
@@ -0,0 +1,280 @@
+
+![Chain of trust](./chain-of-trust.png)
+
+To acquire an SSL certificate you need to prove that you actually own the domain so you can't pretend to be, like, Facebook or something.
+
+So, what does that mean for our internal little site - do we need to prove that we own `badger.facts` so we can get a signed certificate? To do that, we would need to buy the domain name and expose the site to the internet for the CA's to confirm it exists and that we run it. Seems like a lot of work considering the site is only available on our network.
+
+Instead, we can create our own root CA called _Certificates On Me_ (or something) and get all the devices on the network to agree that they trust this CA so that any certificates it signs will be accepted by the browsers.
+
+### Creating your own root CA
+
+There are quite a few steps involved in this, so best if I hand off to an [article I followed](https://deliciousbrains.com/ssl-certificate-authority-for-local-https-development/) to get this configured successfully. This will take you through the steps of:
+
+- configuring your root CA and generating a root CA certificate, which you can do on your Raspberry Pi
+- adding this root CA certificate to the trusted list on your own machine
+- generating certificates signed by this root CA for your website (e.g. `badger.facts`)
+
+Now the SSL certificate you just created, along with any others you create in the future, should be accepted by your machine.
+
+### Bundling certificates
+
+In order to allow a browser to verify the chain of trust from a site's SSL certificate to the (possibly several) intermediate CA that signed it and up to the root CA, it is common for a website to provide a 'bundle' of certificates. This contains all the certificates the browser needs to worry about in one package.
+
+You can create a bundle by inserting the certificates into a single file, starting with the site certificate at the top, followed by the intermediate CA certificates and then the root CA certificate at the bottom.
+
+In our example, we don't have any intermediate CAs so the bundle would contain only 2 certificates.
+
+- `cat badger-facts.crt root_certificate.crt >> badger-facts-bundle.crt` to combine both certificates into a single bundle.
+
+Your certificates may be suffixed with `.pem` to indicate that they are in `PEM` format (bas64 text compared to `DER` binary) - this is a valid format for a `.crt` file so you should be able to copy the contents in without conversion.
+
+### Enabling HTTPS on Nginx
+
+In order to enable HTTPS on our site, we need to tell our web server (which we prepared earlier) to use this new site certificate for all HTTPS requests for our site.
+
+- `sudo nano /etc/nginx/sites-available/default` and add a new `server` section to the bottom of the file that looks like:
+
+```
```

Comment:
Flavour the snippet as nginx 

```nginx
server {
  listen 443;
  server_name _;
  ssl on;
  ssl_certificate /etc/ssl/badger-facts-bundle.crt;
  ssl_certificate_key /etc/ssl/badger-facts.key;
  location / {
    root /var/www/html;
    index index.nginx-debian.html
  }
}
```

here you can see all languages supported  by Github Flavored Markdown https://www.rubycoloredglasses.com/2013/04/languages-supported-by-github-flavored-markdown/